File: common/src/main/java/fishmod/utils/dungeon/map/GridPos.java
package fishmod.utils.dungeon.map;

/**
 * A cell in the dungeon's fixed 6x6 room grid, in half-unit coordinates (0..10): room-quadrant
 * cells sit at even (x, z); door/connector cells sit at odd/even or even/odd. Unlike earlier
 * versions of this class, grid coordinates are no longer derived from the player's world position
 * at all — the whole grid is indexed purely off the map's own pixel data (see MapReader), matching
 * how Odin (github.com/odtheking/Odin) does it. That sidesteps needing any correspondence between
 * world coordinates and map pixel space, which turned out to be the source of miscalibration.
 */
public record GridPos(int x, int z) {
    public boolean isRoomCell() {
        return (x & 1) == 0 && (z & 1) == 0;
    }

    public GridPos offset(int dx, int dz) {
        return new GridPos(x + dx, z + dz);
    }
}

File: common/src/main/kotlin/fishmod/utils/dungeon/map/GridPos.kt
package fishmod.utils.dungeon.map

/** A cell in the dungeon's 6x6 room grid (half-unit coords 0..10, room cells at even x/z), indexed purely from map pixel data (see MapReader), never world position. Plain class (not data class) with explicit x()/z() so Java call sites keep using record-style accessors. */
class GridPos(private val xCoord: Int, private val zCoord: Int) {

    fun x(): Int = xCoord
    fun z(): Int = zCoord

    fun isRoomCell(): Boolean = (xCoord and 1) == 0 && (zCoord and 1) == 0

    fun offset(dx: Int, dz: Int): GridPos = GridPos(xCoord + dx, zCoord + dz)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is GridPos) return false
        return xCoord == other.xCoord && zCoord == other.zCoord
    }

    override fun hashCode(): Int = 31 * xCoord + zCoord

    override fun toString(): String = "GridPos[x=$xCoord, z=$zCoord]"
}
